### PR TITLE
Objdict refactor

### PR DESCRIPTION
**Pull request by cthulahoops**
Refactor into new style declarations.

Now contains changes to vm.to_str and vm.to_repr to make them return PyStringRefs too.

### CODE REVIEW

**Comment by cthulahoops** [comment]
It's easier than I was thinking. I'll do it properly.

File: vm/src/obj/objdict.rs [hunk 1]
```diff
@@ -174,231 +173,146 @@ fn dict_new(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
...
-            str_parts.push(format!("{}: {}", key_str, value_str));
-        }
+    fn repr(self, vm: &mut VirtualMachine) -> PyResult {
+        let s = if let Some(_guard) = ReprGuard::enter(self.as_object()) {
+            let elements = get_key_value_pairs(self.as_object());
+            let mut str_parts = vec![];
+            for (key, value) in elements {
+                let key_repr = vm.to_repr(&key)?;
+                let value_repr = vm.to_repr(&value)?;
+                let key_str = objstr::get_value(&key_repr);
```
Comment (adrian17) [inline vm/src/obj/objdict.rs]: (we probably should change `to_repr()` to return `PyResult<PyStringRef>` - `get_value` calls are a bit ugly and they panic when `__repr__` returns a non-string.)

File: vm/src/obj/objdict.rs [hunk 2]
```diff
@@ -174,231 +173,146 @@ fn dict_new(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
...
+    ) -> PyObjectRef {
+        // What we are looking for:
+        let ref key = key.value;
+
+        let elements = self.entries.borrow();
+        if elements.contains_key(key) {
+            elements[key].1.clone()
+        } else {
+            match default {
+                OptionalArg::Present(ref value) => value.clone(),
```
Comment (adrian17) [inline vm/src/obj/objdict.rs]: you probably don't need to `clone` here:
`OptionalArg::Present(value) => value,`

File: vm/src/obj/objdict.rs (hunk 1 above)
Comment (cthulahoops) [reply vm/src/obj/objdict.rs]: Have changed both ```vm.to_str``` and ```vm.to_repr```.

**Comment by coolreader18** [comment]
I'm not at home so I can't do a proper review, but I saw you used `let ref varname = string.value;`, and IIRC clippy would nag you to change it to `let varname = &string.value;`. Otherwise it looks good!

**Comment by cthulahoops** [comment]
> I'm on mobile so I can't do a proper review, but I saw you used `let ref varname = string.value;`, and IIRC clippy would nag you to change it to `let varname = &string.value;`. Otherwise it looks good!

Thanks. I'll have to get into the habit of using clippy, and break the ```let ref``` habit.

**Pull request merged**